runtime/runtime/src/lib.rs

```diff
@@ -2084,16 +2084,23 @@ impl Runtime {
 
         let bandwidth_requests = receipt_sink.generate_bandwidth_requests(&state_update, true)?;
 
-        check_balance(
-            &apply_state.config,
-            &state_update,
-            validator_accounts_update,
-            processing_state.incoming_receipts,
-            &promise_yield_result.timeout_receipts,
-            processing_state.transactions,
-            &receipt_sink.outgoing_receipts(),
-            &processing_state.stats,
-        )?;
+        if cfg!(debug_assertions) {
+            if let Err(err) = check_balance(
+                &apply_state.config,
+                &state_update,
+                validator_accounts_update,
+                processing_state.incoming_receipts,
+                &promise_yield_result.timeout_receipts,
+                processing_state.transactions,
+                &receipt_sink.outgoing_receipts(),
+                &processing_state.stats,
+            ) {
+                panic!(
+                    "Balance check failed for shard {} at height {} with block hash {}: {}",
```

Comment:
maybe add more context here, because it is not clear to me what you mean by "balance check".